File: test/MATFrostTest/src/MATFrostTest.jl
module MATFrostTest



elementwise_addition_f64(c::Float64, x::Vector{Float64}) = c .+ x

kron_product_matrix_f64(A::Matrix{Float64}, B::Matrix{Float64}) = kron(A, B)

function sum_vector_of_vector_f64(vs::Vector{Vector{Float64}}) :: Float64 
    acc = 0.0
    for v in vs
        for e in v
           acc += e 
        end
    end
    acc
end

struct CompositeNumberType
    v1::Int64
    v2::Int32
    v3::UInt8
end

sum_composite_number_type(cnt::CompositeNumberType) = cnt.v1 + cnt.v2 + cnt.v3

struct SimplePopulationType
    name::String
    population::Int64
end

largest_population(p1::SimplePopulationType, p2::SimplePopulationType) = ifelse(p1.population > p2.population, p1, p2)

function largest_population_vector(ps::Vector{SimplePopulationType})
    mp = ps[1]
    for p in ps
        if mp.population < p.population
            mp = p
        end
    end
    mp
end


struct Nest1
    v1::Float64
    v2::Int64
    v3::Tuple{Float64, Int64, Int32}
    v4::Bool
end

sum_nest1(v::Nest1) :: Float64 = v.v1 + v.v2 + v.v3[1]+v.v3[2]+v.v3[3]

nest1_identity(v::Nest1) = v

struct Nest2
    v1::Tuple{Nest1, Nest1, Nest1}
    vs::Vector{Nest1}
end

# The following function breaks on older Linux Julia versions. Curious behavior.
sum_nest2_breaking(v::Nest2) :: Float64 = sum(sum_nest1(e) for e in v.v1) + sum(sum_nest1(e) for e in v.vs; init=0.0)

function sum_nest2(v::Nest2) :: Float64 
    acc = 0.0
    for e in v.v1
        acc += sum_nest1(e)
    end

    for e in v.vs
        acc += sum_nest1(e)
    end

    acc
end

multiple_method_definitions(x::Float64)="First"
multiple_method_definitions(x::Int64)="Second"
multiple_method_definitions(x::Float64, y::Int64)="Third"


nest2_identity(v::Nest2) = v

identity_nest2(v::Nest2) = v
struct Rename1Old
    old_v  :: Nest1
    old_vs :: Vector{Nest1}
end

struct Rename1New
    action :: String
    new_v  :: Nest1
    new_vs :: Vector{Nest1}
end



rename_test1(v::Rename1Old) = Rename1New("Renamed", v.old_v, v.old_vs)


struct AbstractType1
    v1::Float64
    v2::Any
end


matfrost_abstract_test1(v1::Float64, v2::Int64) = AbstractType1(v1, v2)



matfrost_xor(b1::Bool, b2::Bool) = xor(b1, b2)

matfrost_elementwise_xor(v1::Vector{Bool}, v2::Vector{Bool}) = Vector{Bool}(xor.(v1, v2))


repeat_string(s::String, num::Int64) = reduce(*, (s for _ in 1:num))

concat_strings(s::Vector{String}) = reduce(*, s)


double_scalar_f32(v::Float32) = v+v
double_scalar_f64(v::Float64) = v+v

double_scalar_i8(v::Int8) = v+v
double_scalar_i16(v::Int16) = v+v
double_scalar_i32(v::Int32) = v+v
double_scalar_i64(v::Int64) = v+v


double_scalar_ui8(v::UInt8) = v+v
double_scalar_ui16(v::UInt16) = v+v
double_scalar_ui32(v::UInt32) = v+v
double_scalar_ui64(v::UInt64) = v+v


double_scalar_cf32(v::Complex{Float32}) = v+v
double_scalar_cf64(v::Complex{Float64}) = v+v

double_scalar_ci8(v::Complex{Int8})   = v+v
double_scalar_ci16(v::Complex{Int16}) = v+v
double_scalar_ci32(v::Complex{Int32}) = v+v
double_scalar_ci64(v::Complex{Int64}) = v+v

double_scalar_cui8(v::Complex{UInt8})   = v+v
double_scalar_cui16(v::Complex{UInt16}) = v+v
double_scalar_cui32(v::Complex{UInt32}) = v+v
double_scalar_cui64(v::Complex{UInt64}) = v+v

logical_f64_multiplication(b::Bool, v::Float64) = b * v



double_vector_f32(v::Vector{Float32}) = v .+ v
double_vector_f64(v::Vector{Float64}) = v .+ v

double_vector_i8(v::Vector{Int8})   = v .+ v
double_vector_i16(v::Vector{Int16}) = v .+ v
double_vector_i32(v::Vector{Int32}) = v .+ v
double_vector_i64(v::Vector{Int64}) = v .+ v

double_vector_ui8(v::Vector{UInt8})   = v .+ v
double_vector_ui16(v::Vector{UInt16}) = v .+ v
double_vector_ui32(v::Vector{UInt32}) = v .+ v
double_vector_ui64(v::Vector{UInt64}) = v .+ v



double_vector_cf32(v::Vector{Complex{Float32}}) = v .+ v
double_vector_cf64(v::Vector{Complex{Float64}}) = v .+ v

double_vector_ci8(v::Vector{Complex{Int8}})     = v .+ v
double_vector_ci16(v::Vector{Complex{Int16}})   = v .+ v
double_vector_ci32(v::Vector{Complex{Int32}})   = v .+ v
double_vector_ci64(v::Vector{Complex{Int64}})   = v .+ v

double_vector_cui8(v::Vector{Complex{UInt8}})   = v .+ v
double_vector_cui16(v::Vector{Complex{UInt16}}) = v .+ v
double_vector_cui32(v::Vector{Complex{UInt32}}) = v .+ v
double_vector_cui64(v::Vector{Complex{UInt64}}) = v .+ v




sum_vector_cf64(v::Vector{Complex{Float64}}) = sum(v)




multiplication_scalar_f32(v1::Float32, v2::Float32) = v1*v2
multiplication_scalar_f64(v1::Float64, v2::Float64) = v1*v2

multiplication_scalar_i8(v1::Int8, v2::Int8) = v1*v2
multiplication_scalar_i16(v1::Int16, v2::Int16) = v1*v2
multiplication_scalar_i32(v1::Int32, v2::Int32) = v1*v2
multiplication_scalar_i64(v1::Int64, v2::Int64) = v1*v2


multiplication_scalar_ui8(v1::UInt8, v2::UInt8) = v1*v2
multiplication_scalar_ui16(v1::UInt16, v2::UInt16) = v1*v2
multiplication_scalar_ui32(v1::UInt32, v2::UInt32) = v1*v2
multiplication_scalar_ui64(v1::UInt64, v2::UInt64) = v1*v2


multiplication_scalar_cf32(v1::Complex{Float32}, v2::Complex{Float32}) = v1*v2
multiplication_scalar_cf64(v1::Complex{Float64}, v2::Complex{Float64}) = v1*v2

multiplication_scalar_ci8(v1::Complex{Int8}, v2::Complex{Int8})   = v1*v2
multiplication_scalar_ci16(v1::Complex{Int16}, v2::Complex{Int16}) = v1*v2
multiplication_scalar_ci32(v1::Complex{Int32}, v2::Complex{Int32}) = v1*v2
multiplication_scalar_ci64(v1::Complex{Int64}, v2::Complex{Int64}) = v1*v2

multiplication_scalar_cui8(v1::Complex{UInt8}, v2::Complex{UInt8})   = v1*v2
multiplication_scalar_cui16(v1::Complex{UInt16}, v2::Complex{UInt16}) = v1*v2
multiplication_scalar_cui32(v1::Complex{UInt32}, v2::Complex{UInt32}) = v1*v2
multiplication_scalar_cui64(v1::Complex{UInt64}, v2::Complex{UInt64}) = v1*v2

struct Nest3_L3{T}
    v1::T
    v2::Float64
    v3::Int64
end

struct Nest3_L2{T}
    v1::T
    v2::Vector{T}
    v3::Vector{Nest3_L3{T}}
end

struct Nest3_L1{T}
    v1::Int64
    v2::Float64
    v3::T
    v4::Tuple{T,T,T}
    v5::Vector{T}
    v6::Matrix{T}
    v7::Nest3_L2{T}
    v8::Vector{Nest3_L2{T}}
end

struct Nest4_L1{T}
    v1::Float64
    v2::Int64
    v3::Array{Nest3_L1{T}, 3}
    v4::T
    v5::Nest3_L2{T}
end


function nested_structures_test1(v::Nest3_L1{T}) where {T}
    Nest4_L1{T}(
        v.v2,
        v.v1,
        fill(v, (2, 2, 2)),
        v.v3,
        v.v7
    )
end

function interleave_with_number_and_string(vs::Vector{T}, vnum::Float64, vstring::String) where {T}
    vo = Vector{Union{T, Float64, String}}(undef, length(vs)*3)

    vo[1:3:end] .= vs
    vo[2:3:end] .= vnum
    vo[3:3:end] .= vstring

    vo
end


for (suf, prim) in (
        (:bool, Bool),

        (:string, String),
        
        (:simple_population_type, SimplePopulationType),
        
        (:named_tuple_simple_population_type, @NamedTuple{name::String, population::Int64}),

        (:tup_f64_i64_bool_string, Tuple{Float64, Int64, Bool, String}),




        (:f32, Float32), 
        (:f64, Float64),

        (:i8, Int8), 
        (:ui8, UInt8),
        
        (:i16, Int16), 
        (:ui16, UInt16),
        
        (:i32, Int32), 
        (:ui32, UInt32),
        
        (:i64, Int64), 
        (:ui64, UInt64),




        (:cf32, Complex{Float32}), 
        (:cf64, Complex{Float64}),

        (:ci8, Complex{Int8}), 
        (:cui8, Complex{UInt8}),
        
        (:ci16, Complex{Int16}), 
        (:cui16, Complex{UInt16}),
        
        (:ci32, Complex{Int32}), 
        (:cui32, Complex{UInt32}),
        
        (:ci64, Complex{Int64}), 
        (:cui64, Complex{UInt64}))



    identity = Symbol(:identity_, suf)
    eval(:($(identity)(v::$(prim))=v))

    identity_vector = Symbol(:identity_vector_, suf)
    eval(:($(identity_vector)(v::Vector{$(prim)})=v))

    identity_matrix = Symbol(:identity_matrix_, suf)
    eval(:($(identity_matrix)(v::Matrix{$(prim)})=v))
    
    identity_arr3 = Symbol(:identity_arr3_, suf)
    eval(:($(identity_arr3)(v::Array{$(prim), 3})=v))



    string_s = Symbol(:string_, suf)
    eval(:($(string_s)(v::$(prim)) = string(v)))

    string_vector_s = Symbol(:string_vector_, suf)
    eval(:($(string_vector_s)(v::Vector{$(prim)}) = string(v)))

    string_matrix_s = Symbol(:string_matrix_, suf)
    eval(:($(string_matrix_s)(v::Matrix{$(prim)}) = string(v)))

    string_arr3_s = Symbol(:string_arr3_, suf)
    eval(:($(string_arr3_s)(v::Array{$(prim), 3}) = string(v)))


    eval(:($(Symbol(:nested_structures_test1_scalar_, suf))(v::Nest3_L1{$(prim)}) = nested_structures_test1(v)))

    
    eval(:($(Symbol(:nested_structures_test1_vector_, suf))(v::Nest3_L1{Vector{$(prim)}}) = nested_structures_test1(v)))

    eval(:($(Symbol(:interleave_with_number_and_string_, suf))(vs::Vector{$(prim)}, vnum::Float64, vstring::String) = interleave_with_number_and_string(vs, vnum, vstring)))

    eval(:($(Symbol(:ifelse_, suf))(b::Bool, v::$(prim), vstrings::Vector{String}) = ifelse(b, v, vstrings)))


end







end # module MATFrostTest

File: test/matfrost_test.m
classdef matfrost_test < matfrost_abstract_test
% Unit test for matfrostjulia testing the translations of the base types from MATLAB to Julia and back.

    
    methods(Test)
        % Test methods
        
        function simple_numerical_tests(tc)            
            tc.verifyEqual(...
                tc.mjl.MATFrostTest.elementwise_addition_f64(100, (1:100)'), ...
                100.+(1:100)');

            tc.verifyEqual(...
                tc.mjl.MATFrostTest.kron_product_matrix_f64(diag(1:10), (1:10)'*(1:10)), ...
                kron(diag(1:10), (1:10)'*(1:10)));

        end
% 
%         function sum_vector_of_vector_f64(tc)
%             vov = cell(10,1);
%             acc = 0;
%             for i=1:10
%                 vov{i} = (1:i)';
%                 acc = acc+ sum(1:i);
%             end
% 
%             tc.verifyEqual(...
%                 tc.mjl.MATFrostTest.sum_vector_of_vector_f64(vov), ...
%                 acc);
% 
%         end
        function simple_struct(tc)
            
            p1 = struct;
            p1.name = "Midgard";
            p1.population = int64(5443);

            p2 = struct;
            p2.name = "Asgard";
            p2.population = int64(981);

            
            tc.verifyEqual(...
                tc.mjl.MATFrostTest.largest_population(p1, p2), ...
                p1);
        end

        function simple_struct_vector(tc)
            
            p1 = struct;
            p1.name = "Midgard";
            p1.population = int64(5443);

            p2 = struct;
            p2.name = "Asgard";
            p2.population = int64(981);


            p3 = struct;
            p3.name = "Westeros";
            p3.population = int64(68544);


            p4 = struct;
            p4.name = "Braavos";
            p4.population = int64(23544);

            tc.verifyEqual(...
                tc.mjl.MATFrostTest.largest_population_vector([p1;p2;p3;p4]), ...
                p3);
        end


        function nested_structures_test1(tc)
            
            nest1 = struct;
            nest1.v1 = 3.0;
            nest1.v2 = int64(4);
            nest1.v3 = {5; int64(6); int32(7)};
            nest1.v4 = true;

            tc.verifyEqual(...
                tc.mjl.MATFrostTest.sum_nest1(nest1), ...
                3+4+5+6+7);
        end


        function nested_structures1_identity(tc)

            nest1 = struct;
            nest1.v1 = 3.0;
            nest1.v2 = int64(4);
            nest1.v3 = {5; int64(6); int32(7)};
            nest1.v4 = true;

            tc.verifyEqual(...
                tc.mjl.MATFrostTest.nest1_identity(nest1), ...
                nest1);
        end

        
        function nested_structures_test2(tc)
            nest1 = struct;
            nest1.v1 = 3.0;
            nest1.v2 = int64(4);
            nest1.v3 = {5; int64(6); int32(7)};
            nest1.v4 = true;


            nest2 = struct;
            nest2.v1={nest1;nest1;nest1};
            nest2.vs=repmat(nest1, 50,1);
            
            tc.verifyEqual(...
                tc.mjl.MATFrostTest.sum_nest2(nest2), ...
                (50+3)*(3+4+5+6+7));

        end

        function nested_structures2_identity(tc)
            nest1 = struct;
            nest1.v1 = 3.0;
            nest1.v2 = int64(4);
            nest1.v3 = {5; int64(6); int32(7)};
            nest1.v4 = true;


            nest2 = struct;
            nest2.v1={nest1;nest1;nest1};
            nest2.vs=repmat(nest1, 50,1);
            
            tc.verifyEqual(...
                tc.mjl.MATFrostTest.nest2_identity(nest2), ...
                nest2);

        end
        function rename_structures_test1(tc)
            nest1 = struct;
            nest1.v1 = 3.0;
            nest1.v2 = int64(4);
            nest1.v3 = {5; int64(6); int32(7)};       
            nest1.v4 = true;

            
            rin = struct("old_v", nest1, "old_vs", repmat(nest1, 10,1));

            tc.verifyEqual(...
                tc.mjl.MATFrostTest.rename_test1(rin), ...
                struct( ...
                    "action", "Renamed", ...
                    "new_v", nest1, ...
                    "new_vs", repmat(nest1, 10,1) ...
                ));

        end

        function rename_structures_test2(tc)
            nest1 = struct;
            nest1.v1 = 3.0;
            nest1.v2 = int64(4);
            nest1.v3 = {5; int64(6); int32(7)};       
            nest1.v4 = true;

            
            rin = struct("old_v", nest1, "old_vs", repmat(nest1, 10,1));

            tc.verifyEqual(...
                tc.mjl.MATFrostTest.rename_test1(rin), ...
                struct( ...
                    "action", "Renamed", ...
                    "new_v", nest1, ...
                    "new_vs", repmat(nest1, 10,1) ...
                ));

        end

        function abstract_type_test1(tc)
            
            
            
            tc.verifyEqual(...
                tc.mjl.MATFrostTest.matfrost_abstract_test1(1.3, int64(4)), ...
                struct( ...
                    "v1", 1.3, ...
                    "v2", int64(4)...
                ));

        end
        

        function string_manipulations(tc)
            
            tc.verifyEqual(...
                tc.mjl.MATFrostTest.repeat_string("teststring_😀_", int64(4)), ...
                "teststring_😀_teststring_😀_teststring_😀_teststring_😀_");


        end
        
        function vector_of_strings_concatanation(tc)
            
            tc.verifyEqual(...
                tc.mjl.MATFrostTest.concat_strings(["teststring_😀_"; "_lorem_ipusm_"; "_teststring2_🙁"]), ...
                    "teststring_😀__lorem_ipusm__teststring2_🙁");

        end


        function primitive_logical(tc)
            tc.verifyEqual(...
                    tc.mjl.MATFrostTest.logical_f64_multiplication(true, 8.8), ...
                    8.8);
            tc.verifyEqual(...
                    tc.mjl.MATFrostTest.logical_f64_multiplication(false, 8.8), ...
                    0.0)
        end

        function double_primitive_scalars(tc)
            
            ts = ["i8"; "ui8"; "i16"; "ui16"; "i32"; "ui32"; "i64"; "ui64"; ...
                "f32"; "f64"];

            vs = {...
                int8(22);         uint8(42);...
                int16(3322);      uint16(4532);...
                int32(323442);    uint32(53342);... 
                int64(323434542); uint64(535345342);...
                ...
                single(34.125);   2342.0625};

            % Real numbers
            for i=1:numel(ts)
                tc.verifyEqual(...
                    tc.mjl.MATFrostTest.("double_scalar_" + ts(i))(vs{i}), vs{i}+vs{i});
            end
        end

        function double_complex_primitive_scalars(tc)
            ts = ["i8"; "ui8"; "i16"; "ui16"; "i32"; "ui32"; "i64"; "ui64"; ...
                "f32"; "f64"];

            vs = {...
                int8(22);         uint8(42);...
                int16(3322);      uint16(4532);...
                int32(323442);    uint32(53342);... 
                int64(323434542); uint64(535345342);...
                ...
                single(34.125);   2342.0625};


            % Complex numbers
            for i=1:numel(ts)
                v  = complex(vs{i},   2 * vs{i});
                v2 = complex(2*vs{i}, 4 * vs{i});
                tc.verifyEqual(...
                    tc.mjl.MATFrostTest.("double_scalar_c" + ts(i))(v), v2);
            end

        end

        function double_primitive_vectors(tc)
            
            ts = ["i8"; "ui8"; "i16"; "ui16"; "i32"; "ui32"; "i64"; "ui64"; ...
                "f32"; "f64"];

            vs = {...
                int8(11 + (1:10)');         uint8(42 + (1:10)');...
                int16(3322 + (1:10)');      uint16(4532 + (1:10)');...
                int32(323442 + (1:10)');    uint32(53342 + (1:10)');... 
                int64(323434542 + (1:10)'); uint64(535345342 + (1:10)');...
                ...
                single(34.125 + (1:10)');   2342.0625 + (1:10)'};

            % Real numbers
            for i=1:numel(ts)
                tc.verifyEqual(...
                    tc.mjl.MATFrostTest.("double_vector_" + ts(i))(vs{i}), vs{i}+vs{i});
            end
            
        end

        
        function empty_vectors(tc)
            
            ts = ["i8"; "ui8"; "i16"; "ui16"; "i32"; "ui32"; "i64"; "ui64"; ...
                "f32"; "f64"];

            % Real numbers
            for i=1:numel(ts)
                tc.verifyEqual(...
                    tc.mjl.MATFrostTest.("double_vector_" + ts(i))([]), []);
                    
            end
            
        end

        function double_complex_primitive_vectors(tc)
            
            ts = ["i8"; "ui8"; "i16"; "ui16"; "i32"; "ui32"; "i64"; "ui64"; ...
                "f32"; "f64"];

            vs = {...
                int8(11 + (1:10)');         uint8(42 + (1:10)');...
                int16(3322 + (1:10)');      uint16(4532 + (1:10)');...
                int32(323442 + (1:10)');    uint32(53342 + (1:10)');... 
                int64(323434542 + (1:10)'); uint64(535345342 + (1:10)');...
                ...
                single(34.125 + (1:10)');   2342.0625 + (1:10)'};

            % Complex numbers
            for i=1:numel(ts)
                v  = complex(vs{i},   2 * vs{i});
                v2 = complex(2*vs{i}, 4 * vs{i});
                tc.verifyEqual(...
                    tc.mjl.MATFrostTest.("double_vector_c" + ts(i))(v), v2);
            end

        end


        function sum_complex_numbers(tc)
            vr = 2342.0625 + logspace(1,5,100)';
            vi = 345.0625  + logspace(3,4,100)';
            
            v = complex(vr,vi);
            tc.verifyEqual(...
                    tc.mjl.MATFrostTest.("sum_vector_cf64")(v), complex(sum(vr), sum(vi)), "RelTol", 1e-15);
        end

    end

    
end
